src/quic/streams.cc

```diff
@@ -1,61 +1,1052 @@
+
+#define STREAM_STATS(V)                                                        \
+  V(CREATED_AT, created_at)                                                    \
+  V(RECEIVED_AT, received_at)                                                  \
+  V(ACKED_AT, acked_at)                                                        \
+  V(CLOSING_AT, closing_at)                                                    \
+  V(DESTROYED_AT, destroyed_at)                                                \
+  V(BYTES_RECEIVED, bytes_received)                                            \
+  V(BYTES_SENT, bytes_sent)                                                    \
+  V(MAX_OFFSET, max_offset)                                                    \
+  V(MAX_OFFSET_ACK, max_offset_ack)                                            \
+  V(MAX_OFFSET_RECV, max_offset_received)                                      \
+  V(FINAL_SIZE, final_size)
+
+#define STREAM_JS_METHODS(V)                                                   \
+  V(AttachSource, attachSource, false)                                         \
+  V(Destroy, destroy, false)                                                   \
+  V(SendHeaders, sendHeaders, false)                                           \
+  V(StopSending, stopSending, false)                                           \
+  V(ResetStream, resetStream, false)                                           \
+  V(SetPriority, setPriority, false)                                           \
+  V(GetPriority, getPriority, true)                                            \
+  V(GetReader, getReader, false)
+
+struct Stream::State {
+#define V(_, name, type) type name;
+  STREAM_STATE(V)
+#undef V
+};
+
+STAT_STRUCT(Stream, STREAM)
+
+// ============================================================================
+
+namespace {
+Maybe<std::shared_ptr<DataQueue>> GetDataQueueFromSource(Environment* env,
+                                                         Local<Value> value) {
+  DCHECK_IMPLIES(!value->IsUndefined(), value->IsObject());
+  if (value->IsUndefined()) {
+    return Just(std::shared_ptr<DataQueue>());
+  } else if (value->IsArrayBuffer()) {
+    auto buffer = value.As<ArrayBuffer>();
+    std::vector<std::unique_ptr<DataQueue::Entry>> entries(1);
+    entries.push_back(DataQueue::CreateInMemoryEntryFromBackingStore(
+        buffer->GetBackingStore(), 0, buffer->ByteLength()));
+    return Just(DataQueue::CreateIdempotent(std::move(entries)));
+  } else if (value->IsSharedArrayBuffer()) {
+    auto buffer = value.As<SharedArrayBuffer>();
+    std::vector<std::unique_ptr<DataQueue::Entry>> entries(1);
+    entries.push_back(DataQueue::CreateInMemoryEntryFromBackingStore(
+        buffer->GetBackingStore(), 0, buffer->ByteLength()));
+    return Just(DataQueue::CreateIdempotent(std::move(entries)));
+  } else if (value->IsArrayBufferView()) {
+    std::vector<std::unique_ptr<DataQueue::Entry>> entries(1);
```

Comment:
Do we need a `std::vector` here? Can we make it a `std::array` since it is fixed length? (Same applies to previous std::vector as well)

Reply:
Perhaps as a separate change. The dataqueue/queue.h api expects std::vector on input (keeping in mind that these are used for Blobs which have arbitrarily large numbers of inputs). I'll keep this as is for now.